aws/resource_aws_dx_hosted_private_virtual_interface_test.go

```diff
@@ -74,42 +157,62 @@ func testAccCheckAwsDxHostedPrivateVirtualInterfaceDestroy(s *terraform.State) e
 		if rs.Type != "aws_dx_hosted_private_virtual_interface" {
 			continue
 		}
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("No ID is set")
+		}
 
-		input := &directconnect.DescribeVirtualInterfacesInput{
+		resp, err := conn.DescribeVirtualInterfaces(&directconnect.DescribeVirtualInterfacesInput{
 			VirtualInterfaceId: aws.String(rs.Primary.ID),
+		})
+		if isAWSErr(err, directconnect.ErrCodeClientException, "does not exist") {
+			continue
 		}
-
-		resp, err := conn.DescribeVirtualInterfaces(input)
 		if err != nil {
 			return err
 		}
-		for _, v := range resp.VirtualInterfaces {
-			if *v.VirtualInterfaceId == rs.Primary.ID && !(*v.VirtualInterfaceState == directconnect.VirtualInterfaceStateDeleted) {
-				return fmt.Errorf("[DESTROY ERROR] Dx Private VIF (%s) not deleted", rs.Primary.ID)
+
+		n := len(resp.VirtualInterfaces)
+		switch n {
+		case 0:
+			continue
+		case 1:
+			if aws.StringValue(resp.VirtualInterfaces[0].VirtualInterfaceState) == directconnect.VirtualInterfaceStateDeleted {
+				continue
 			}
+			return fmt.Errorf("still exist.")
+		default:
+			return fmt.Errorf("Found %d Direct Connect virtual interfaces for %s, expected 1", n, rs.Primary.ID)
 		}
 	}
+
 	return nil
 }
 
-func testAccCheckAwsDxHostedPrivateVirtualInterfaceExists(name string) resource.TestCheckFunc {
+func testAccCheckAwsDxHostedPrivateVirtualInterfaceExists(name string, vif *directconnect.VirtualInterface) resource.TestCheckFunc {
 	return func(s *terraform.State) error {
-		_, ok := s.RootModule().Resources[name]
+		conn := testAccProvider.Meta().(*AWSClient).dxconn
+
+		rs, ok := s.RootModule().Resources[name]
 		if !ok {
 			return fmt.Errorf("Not found: %s", name)
 		}
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("No ID is set")
+		}
 
-		return nil
-	}
-}
+		resp, err := conn.DescribeVirtualInterfaces(&directconnect.DescribeVirtualInterfacesInput{
+			VirtualInterfaceId: aws.String(rs.Primary.ID),
+		})
+		if err != nil {
+			return err
+		}
 
-func testAccCheckAwsDxHostedPrivateVirtualInterfaceAccepterExists(name string) resource.TestCheckFunc {
-	return func(s *terraform.State) error {
-		_, ok := s.RootModule().Resources[name]
-		if !ok {
-			return fmt.Errorf("Not found: %s", name)
+		if n := len(resp.VirtualInterfaces); n != 1 {
```

Comment:
Same note about using `for` loop instead applies here. Generally when we start seeing this logic repeated, its easiest to refactor it into its own function. 👍

Reply:
Same as [above](https://github.com/terraform-providers/terraform-provider-aws/pull/9572/files#r310260719).

Reply:
See comment [above](https://github.com/terraform-providers/terraform-provider-aws/pull/9572#discussion_r310261371).